pkg/resources/tag_masking_policy_association.go

```diff
@@ -102,98 +85,70 @@ func TagMaskingPolicyAssociation() *schema.Resource {
+	if err := client.Tags.Alter(ctx, sdk.NewAlterTagRequest(tid).WithSet(set)); err != nil {
+		return diag.FromErr(err)
+	}
+	aid := attachmentID{
+		TagDatabaseName:           tid.DatabaseName(),
+		TagSchemaName:             tid.SchemaName(),
+		TagName:                   tid.Name(),
+		MaskingPolicyDatabaseName: mid.DatabaseName(),
+		MaskingPolicySchemaName:   mid.SchemaName(),
+		MaskingPolicyName:         mid.Name(),
+	}
+	fmt.Printf("attachment id: %s\n", aid.String())
+	d.SetId(aid.String())
+	return ReadContextTagMaskingPolicyAssociation(ctx, d, meta)
 }
 
-// ReadTagTagMaskingPolicyAssociation implements schema.ReadFunc.
-func ReadTagMaskingPolicyAssociation(d *schema.ResourceData, meta interface{}) error {
+func ReadContextTagMaskingPolicyAssociation(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	diags := diag.Diagnostics{}
 	db := meta.(*sql.DB)
-	attachementID, err := attachedPolicyIDFromString(d.Id())
+	client := sdk.NewClientFromDB(db)
+
+	aid, err := parseAttachmentID(d.Id())
 	if err != nil {
-		return err
+		return diag.FromErr(err)
 	}
 
-	tagDBName := attachementID.TagDatabaseName
-	tagSchemaName := attachementID.TagSchemaName
-	tagName := attachementID.TagName
-	mpDBName := attachementID.MaskingPolicyDatabaseName
-	mpSchameName := attachementID.MaskingPolicySchemaName
-	mpName := attachementID.MaskingPolicyName
-
-	mP := snowflake.MaskingPolicy(mpName, mpDBName, mpSchameName)
-	builder := snowflake.NewTagBuilder(tagName).WithDB(tagDBName).WithSchema(tagSchemaName).WithMaskingPolicy(mP)
-
 	// create temp warehouse to query the tag, and make sure to clean it up
-	client := sdk.NewClientFromDB(db)
-	ctx := context.Background()
-	originalWarehouse, err := client.ContextFunctions.CurrentWarehouse(ctx)
+	warehouse, err := client.ContextFunctions.CurrentWarehouse(ctx)
 	if err != nil {
-		return err
+		return diag.FromErr(err)
 	}
-	if originalWarehouse == "" {
+	if warehouse == "" {
 		log.Printf("[DEBUG] no current warehouse set, creating a temporary warehouse")
 		randomWarehouseName := fmt.Sprintf("terraform-provider-snowflake-%v", helpers.RandomString())
-		tempWarehouseID := sdk.NewAccountObjectIdentifier(randomWarehouseName)
-		err = client.Warehouses.Create(ctx, tempWarehouseID, nil)
-		if err != nil {
-			return err
+		wid := sdk.NewAccountObjectIdentifier(randomWarehouseName)
+		if err := client.Warehouses.Create(ctx, wid, nil); err != nil {
+			return diag.FromErr(err)
 		}
 		defer func() {
-			err := client.Warehouses.Drop(ctx, tempWarehouseID, nil)
-			if err != nil {
+			if err := client.Warehouses.Drop(ctx, wid, nil); err != nil {
 				log.Printf("[WARN] error cleaning up temp warehouse %v", err)
 			}
-			err = client.Sessions.UseWarehouse(ctx, sdk.NewAccountObjectIdentifier(originalWarehouse))
-			if err != nil {
+			if err := client.Sessions.UseWarehouse(ctx, sdk.NewAccountObjectIdentifier(warehouse)); err != nil {
```

Comment:
this will always throw an error because this will only happen when `warehouse == ""`

Reply:
removed. when wh is dropped, it will remove current warehouse anyways